adr/graph-feature-flags.md

```diff
@@ -0,0 +1,212 @@
+wgc subgraph publish posts --schema ./subgraph-posts/schema.graphql
+```
+
+2. Create and publish a users subgraph
+
+```shell
+wgc subgraph create users --label team=B --routing-url https://users.domain.com
+wgc subgraph publish users --schema ./subgraph-users/schema.graphql
+```
+
+3. Create a federated Graph
+
+```shell
+wgc federated-graph create production --label-matcher team=A,team=B --routing-url https://graph.domain.com/graphql
+```
+
+4. Create a feature flag for the users subgraph
+
+```shell
+wgc feature-flag create users-v2 --label team=B --routing-url https://users-v2.domain.com --subgraph users
+# alias: wgc ff create
+wgc feature-flag publish users-v2 --schema ./subgraph-users-v2/schema.graphql
+# alias: wgc ff publish
+```
+
+5. make a request to the federated graph with the feature flag
+
+```shell
+curl -X POST https://graph.domain.com/graphql \
+  -H "Content-Type: application/json" \
+  -H "X-Feature-Flag: users-v2"
+  -d '{"query": "{ users { id name } }"}'
+```
+
+Feature flags can also be enabled via cookies:
+
+```shell
+curl -X POST https://graph.domain.com/graphql \
+  -H "Content-Type: application/json" \
+  -H "Cookie: feature-flag=users-v2"
+  -d '{"query": "{ users { id name } }"}'
+```
+
+Feature flags can also be enabled via "extensions" field in the GraphQL request:
+
+```shell
+curl -X POST https://graph.domain.com/graphql \
+  -H "Content-Type: application/json" \
+  -d '{"query": "{ users { id name } }", "extensions": { "featureFlag": "users-v2" } }'
+```
+
+6. Create a second feature flag for the posts subgraph
+
+```shell
+wgc ff create posts-v2 --label team=A --routing-url https://posts-v2.domain.com --subgraph posts
+wgc ff publish posts-v2 --schema ./subgraph-posts-v2/schema.graphql
+```
+
+7. Create a group feature flag
+
+```shell
+wgc feature-flag-group create v2 --label team=A --flags users-v2,posts-v2
+# alias: wgc ffg create
+```
+
+8. make a request to the federated graph with the group feature flag
+
+This will enable both the users-v2 and posts-v2 feature flags
+
+```shell
+curl -X POST https://graph.domain.com/graphql \
+  -H "Content-Type: application/json" \
+  -H "X-Feature-Flag: v2"
+  -d '{"query": "{ users { id name } posts { id title } }"}'
+```
+
+9. Delete the group feature flag and both feature flags
+
+```shell
+wgc ffg delete v2
+wgc ff delete users-v2
+wgc ff delete posts-v2
+```
+
+## Context
+
+### Feature Flags always replace the subgraph
+
+Feature Flags are always replacing the subgraph they are associated with.
+This means that a feature flag can only be associated with one subgraph at a time.
+If a feature flag is associated with a subgraph, it will replace the subgraph in the federated graph.
+
+### Conflicting Feature Flags that replace the same subgraph are allowed
+
+It's possible to have multiple feature flags that replace the same subgraph.
+This is useful and allowed for testing different variants of the same subgraph.
```

Comment:
Does this mean then we can reference multiple flags for the same subgraph in the same request?

Reply:
No, you can only have one feature flag at a time per Subgraph.
Does this work for your use case?

Reply:
One feature flag at a time per subgraph does meet the use case. My follow up question is what would happen if there are multiple flags for the same subgraph i.e someone forgot to remove a cookie for the other feature flag?